content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+  - Without refactoring the design of the program will decay over time.
+- Refactoring makes software easier to understand.
+  - *“Any fool can write code that a computer can understand. Good programmers write code that humans can understand.”* - Martin Fowler
+- Refactoring makes you find bugs.
+  - When the code is more simple, it is easier to understand, easier to debug, and also easier to detect non-considered cases.
+- Refactoring helps you program faster.
+  - Without a good design, you can progress quickly for a while, but soon the poor design starts to slow you down.
+
+**Conclusion: Refactoring lowers the cost of enhancements.**
+
+## 2) WHEN TO DO REFACTORING?
+Refactoring isn't a special task that will show up in a project plan, so when should you do it?
+- When you need to add a new feature to a codebase, look at the existing code and consider whether it's structured in such a way to make the new change straightforward.
+  - If it isn't, then refactor the existing code to make this new addition easy.
+  - Once that change is done, then add the new feature.
+- After you add a feature and get it working, you can review the resulting code, that while it works, it may not be as clear as it could be.
+  - If it isn't, then refactor it into better shape so that when you (or someone else) return to this code in a few weeks, won't have to spend time puzzling out how this code works.
+- If you struggle to understand the code, refactor it so you won't have to struggle again next time you look at it.
+
+**Conclusion: Refactoring is a part of day-to-day programming.**
+
+## 3) CODE SMELLS
+Probably you’ve heard about code smells, and for sure you have already seen tons of them. Let’s see their characteristics.
+- A code smell is a surface indication that usually corresponds to a deeper problem in the system.
+  - Firstly a smell is by definition something quick to spot - or sniffable.
+  - A long method is a good example of this - just looking at the code and my nose twitches if I see more than a dozen lines.
+- The second is that smells don't always indicate a problem.
+  - Some long methods are just fine. You have to look deeper to see if there is an underlying problem - smells aren't inherently bad on their own - they are often an indicator of a problem rather than the problem themselves.
+
+See the example below (it’s written in Ruby but should be simple enough for anyone). I know the white text boxes already grab much attention, just cover them with your hand and look at the code for no more than 10 seconds, thinking of the code smells that you can find. Then uncover the white boxes. I’m quite sure that in that amount of time you could find the same code smells (or even more). What I’m trying to show, is exactly what code smells are, they are something very quick to spot, and they are not a problem by themselves (think that this code works perfectly fine).
```

Comment:
1. "Below is an example written in Ruby (it should be simple enough for anyone to follow). Don't look at the image yet!"
2. "I know the grey boxes..."